sssd_test_framework/utils/adcli.py

```diff
@@ -0,0 +1,198 @@
+
+        self.cli: CLIBuilder = self.host.cli
+        """Command line builder."""
+
+        self.fs: LinuxFileSystem = fs
+        """Filesystem utils."""
+
+    def _info(
+        self,
+        *,
+        domain: str | None = None,
+        domain_controller: str | None,
+    ) -> str:
+        """
+        Call ``adcli info `` with given arguments.
+
+        :param domain: Displays discovered information about an Active Directory domain, defaults to None
+        :type domain: str | None,
+        :param domain_controller: Domain controller to connect, defaults to None
+        :type domain_controller: str | None, optional
+        """
+        args: CLIBuilderArgs = {
+            "domain": (self.cli.option.VALUE, domain),
+            "domain_controller": (self.cli.option.VALUE, domain_controller),
+        }
+
+        self.host.conn.exec(["adcli", "info"] + self.cli.args(args))
+
+    def _testjoin(
+        self,
+        *,
+        domain: str| None = None,
+        domain_controller: str| None = None,
+        host_keytab: str| None = None,
+    ) -> str:
+        """
+        call ``adcli testjoin `` with given arguments.
+
+        :param domain: Target Active Directory domain, defaults to None
+        :type domain: str | None, optional
+        :param domain_controller: Domain controller to connect
+        :type domain_controller: str | None, optional
+        """
+        args: CLIBuilderArgs = {
+            "domain": (self.cli.option.VALUE, domain),
+            "domain_controller": (self.cli.option.VALUE, domain_controller),
+            "host_keytab": (self.cli.option.VALUE, host_keytab),
+        }
+        self.host.conn.exec(["adcli", "testjoin"] + self.cli.args(args))
+
+    def _join(
+        self,
+        *,
+        domain: str| None = None,
+        domain_controller: str| None = None,
+        domain_realm: str| None = None,
+        host_fqdn: str| None = None,
+        host_keytab: str| None = None,
+        computer_name: str| None = None,
+        login_ccache: str| None = None,
+        login_user: str| None = None,
+        login_type: str| None = None,
+        domain_ou: str| None = None,
+        service_name: str| None = None,
+        os_name: str| None = None,
+        os_version: str| None = None,
+        os_service_pack: str| None = None,
+        user_principal: str| None = None,
+        trusted_for_delegation: str| None = None,
+        dont_expire_password: str| None = None,
+        add_service_principal: str| None = None,
+        description: str| None = None,
+        setattrs: str| None = None,
+        no_password: str| None = None,
+        promp_password: str| None = None,
+        stdin_password: str| None = None,
+        one_time_password: str| None = None,
+        show_password: str| None = None,
+        show_details: str| None = None,
+        add_samba_data: str| None = None,
+        samba_data_tool: str| None = None,
+        ldap_passwd: str| None = None,
```

Comment:
Use **kwargs or *args to handle additional argument and keep only required arguments here. Otherwise it would be difficult to maintain.

Reply:
ack

Reply:
implemented.